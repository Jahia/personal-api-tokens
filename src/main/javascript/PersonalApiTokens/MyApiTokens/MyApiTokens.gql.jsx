import gql from 'graphql-tag';

const CreateTokenMutation = gql`
    mutation CreateToken($userId: String!, $name: String!, $expireAt: String!) {
        admin {
             personalApiTokens {
                 createToken(userId: $userId, name: $name, expireAt: $expireAt)
            }
        }
    }
`;

const getTokenData = gql`
    query GetTokenData($token: String!) {
          admin {
            personalApiTokens {
             token(token: $token) {
               key
               name
               user {
                 name
               }
             }
            }
          }
        }
`;

const getCurrentUserName = gql`
    {
      currentUser {
        name
      }
    }
`;

export {CreateTokenMutation, getTokenData, getCurrentUserName};
