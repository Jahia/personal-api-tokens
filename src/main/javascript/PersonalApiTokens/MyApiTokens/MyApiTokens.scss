.root {
    display: flex;
    flex: 1;
    flex-direction: column;
    padding: 0;
    overflow-x: hidden;
    overflow-y: auto;

    background-color: var(--color-gray_light40);
    box-shadow: 0 4px 4px rgba(59, 61, 64, 0.3);
}

.headerRoot {
    padding: var(--spacing-medium) var(--spacing-large) 0 var(--spacing-large);

    background-color: var(--color-white);
    box-shadow: 0 4px 8px var(--color-dark40);
}

.header {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-start;
    min-height: 48px;
}

.content {
    display: flex;
    flex: 1;
    padding: var(--spacing-medium) var(--spacing-large) var(--spacing-medium) var(--spacing-large);

    background-color: var(--color-gray_light40);
}

.actionBar {
    display: flex;
    flex-grow: 0.05;
    flex-direction: row-reverse;
    justify-content: space-between;
    align-items: center;
}
