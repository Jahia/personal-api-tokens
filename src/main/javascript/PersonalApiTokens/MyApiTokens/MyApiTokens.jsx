import React, {useState} from 'react';
import {useTranslation} from 'react-i18next';
import {Add, Button, GlobalStyle, Typography} from '@jahia/moonstone';
import TokensList from '../TokensList/TokensList';
import styles from './MyApiTokens.scss';
import ConfirmationDialog from '../ConfirmationDialog/ConfirmationDialog';
import CreateTokenDialogBody from '../CreateTokenDialogBody/CreateTokenDialogBody';
import {MuiPickersUtilsProvider} from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';
import moment from 'moment';
import CopyTokenDialogBody from '../CopyTokenDialogBody/CopyTokenDialogBody';
import {CreateTokenMutation, getUserInformation} from './MyApiTokens.gql';
import {useMutation, useQuery} from '@apollo/react-hooks';
import {useLocation} from 'react-router-dom';

const MyApiTokens = () => {
    const {t} = useTranslation('personal-api-tokens');
    const location = useLocation();
    const user = new URLSearchParams(location.search).get('user');

    const [isCreateTokenDialogOpen, setCreateTokenDialogOpen] = useState(false);
    const [isCopyTokenDialogOpen, setCopyTokenDialogOpen] = useState(false);
    const [createTokenError, setCreateTokenError] = useState(false);
    const [userTokenInformation, setUserTokenInformation] = useState({userId: user ? user : contextJsParameters.user.path, name: '', expireAt: moment().add(1, 'days')});
    const [tokenValue, setTokenValue] = useState('');

    // This is probably gonna be used
    // eslint-disable-next-line
    const updateCurrentUser = data => {
        setUserTokenInformation({...userTokenInformation, userId: data.currentUser.name});
    };

    const refreshState = () => {
        setCreateTokenError(false);
        setUserTokenInformation({...userTokenInformation, name: '', expireAt: moment().add(1, 'days')});
        setTokenValue('');
    };

    const userInformation = useQuery(getUserInformation, {
        variables: {
            userPath: userTokenInformation.userId
        }
    });

    const updateTokenValue = data => {
        setCreateTokenDialogOpen(false);
        setCopyTokenDialogOpen(true);
        const tokenData = data.admin.personalApiTokens.createToken;
        setTokenValue(tokenData ? tokenData : '');
    };

    const [createTokenMutation] = useMutation(CreateTokenMutation, {
        onCompleted: updateTokenValue,
        onError: () => setCreateTokenError(true),
        variables: userTokenInformation
    });

    return (
        <MuiPickersUtilsProvider utils={MomentUtils}>
            <GlobalStyle/>
            <div className={styles.root}>
                <div className={styles.headerRoot}>
                    <header className={styles.header}>
                        <Typography variant="title">
                            {user ? t('personal-api-tokens:adminTitle', {name: userInformation?.data?.jcr?.nodeByPath?.displayName}) : t('personal-api-tokens:title')}
                        </Typography>
                        <div className={styles.actionBar}>
                            <Button size="big"
                                    color="accent"
                                    label={t('personal-api-tokens:createToken.buttonTitle')}
                                    icon={<Add/>}
                                    onClick={() => {
                                        refreshState();
                                        setCreateTokenDialogOpen(true);
                                    }}/>
                        </div>
                    </header>
                </div>
                <div className={styles.content}>
                    <ConfirmationDialog isOpen={isCreateTokenDialogOpen}
                                        acceptLabel={t('personal-api-tokens:create')}
                                        cancelLabel={t('personal-api-tokens:cancel')}
                                        title={t('personal-api-tokens:createToken.modalTitle')}
                                        body={<CreateTokenDialogBody
                                            tokenInformation={userTokenInformation}
                                            setTokenInformation={setUserTokenInformation}
                                            error={createTokenError}
                                        />}
                                        validationState={{acceptButtonDisabled: userTokenInformation.name === ''}}
                                        onClose={() => setCreateTokenDialogOpen(false)}
                                        onAccept={() => createTokenMutation()}/>
                    <ConfirmationDialog isOpen={isCopyTokenDialogOpen}
                                        acceptLabel={t('personal-api-tokens:close')}
                                        title={t('personal-api-tokens:copyToken.title')}
                                        body={<CopyTokenDialogBody tokenValue={tokenValue}/>}
                                        onClose={() => setCopyTokenDialogOpen(false)}
                                        onAccept={() => setCopyTokenDialogOpen(false)}/>
                    <TokensList/>
                </div>
            </div>
        </MuiPickersUtilsProvider>
    );
};

export default MyApiTokens;
