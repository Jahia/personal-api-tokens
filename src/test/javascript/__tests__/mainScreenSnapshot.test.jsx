import React from 'react';
import {act, cleanup, render} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import MyApiTokens from '../../../main/javascript/PersonalApiTokens/MyApiTokens/MyApiTokens';
import {snapshotMocks} from '../apolloMocks';
import {MockedProvider, wait} from '@apollo/react-testing';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: () => ({
        pathname: 'localhost:3000/example/path'
    })
}));

global.contextJsParameters = {
    user: {
        username: 'root'
    }
};

describe('Validate main screen snapshot', () => {
    afterEach(() => {
        cleanup();
    });
    test('match snapshot', async () => {
        const {container} = render(
            <MockedProvider mocks={snapshotMocks} addTypename={false}>
                <MyApiTokens/>
            </MockedProvider>
        );
        await act(async () => {
            await wait(0);
        });
        expect(container).toMatchSnapshot();
    });
});

