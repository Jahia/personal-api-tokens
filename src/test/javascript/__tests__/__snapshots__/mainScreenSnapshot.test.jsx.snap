// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate main screen snapshot match snapshot 1`] = `
<div>
  <main
    class="moonstone-contentLayout flexCol"
  >
    <header
      class="moonstone-contentHeader"
    >
      <div
        class="moonstone-contentHeader_upper"
      >
        <div
          class="flexRow"
        >
          <p
            class="moonstone-typography moonstone-variant_title moonstone-weight_default"
          >
            translated_personal-api-tokens:title
          </p>
          <div
            class="flexFluid"
          />
          <button
            class="moonstone-button moonstone-size_big moonstone-variant_default moonstone-color_accent moonstone-icon"
            data-testid="create-token-btn"
            type="button"
          >
            <svg
              class="undefined moonstone-icon moonstone-icon_default"
              fill="currentColor"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
            <span
              class="moonstone-typography moonstone-variant_button moonstone-weight_semiBold flexFluid moonstone-nowrap moonstone-upperCase"
            >
              translated_personal-api-tokens:createToken.buttonTitle
            </span>
          </button>
        </div>
      </div>
    </header>
    <div
      class="moonstone-contentLayout_paperContainer flexCol"
    >
      <div
        class="moonstone-contentLayout_paper flexCol"
      >
        <div
          class="tokensList flexRow_center alignCenter"
        >
          <p
            class="moonstone-typography moonstone-variant_subheading moonstone-weight_light"
            data-testid="no-tokens-message"
          >
            translated_personal-api-tokens:noTokens
          </p>
        </div>
      </div>
    </div>
  </main>
</div>
`;
